### ISSUE TITLE
Integrate with restic/kopia for base backup de-duplication

### ISSUE THREAD

**Issue by haslersn** [wal-g/wal-g#1792]
### Database name

PostgreSQL

# Issue description

When taking daily base backups, after some time they take up a lot of storage. E.g., for some of our DBs, each LZMA-compressed base backup has a size of more than 10 GB. As we didn't delete backups for two years, for a single DB, more than 7 TB of base backups have accumulated, which is much more than the size of WAL files during the same time period.

Of course space can be freed by deleting old backups or keeping, e.g., only one monthly base backup after some time. However, I think this problem can be solved without deleting old backups, using inter-backup de-duplication. Note that most of the content of a base backup doesn't change often.

## Proposed solution

wal-g could integrate with existing de-duplication solutions such as kopia/restic (which use [content defined chunking (CDC)](https://restic.net/blog/2015-09-12/restic-foundation1-cdc/)) for writing and reading base backups. For WAL files, de-duplication probably wouldn't have much effect. In this case, the wal-g repository would have a subdirectory containing the kopia/restic repository (for storing base backups) and a wal_005 subdirectory (as currently, for WAL files).

**Comment by x4m** [comment]
Hi! Do you know we have delta backups? We deduplicate on LSN basis, 8kb pages that did not change will not be backed up.

**Comment by haslersn** [comment]
Actually I didn't know about that. Now that you pointed me to it, I see that the wal-g docs mentions delta backups at some places, but there doesn't seem to be any introduction or guide to it in the docs.

I assume that one typically still wants to have a full base backup from time to time, in order to
- keep restore time within reasonable bounds and
- be able to delete older backups.

Then these full base backups would still accumulate (although slower) without de-duplication between them.

In the short-term, we will probably solve this problem by writing wal-g backups to a local filesystem, where we retain only the most recent wal-g backup and then take kopia backups from that filesystem. In this case, would there be any advantage to using delta backups? E.g., are delta base backups significantly faster than full base backups?

**Comment by x4m** [comment]
Yup, deltas are order of magnitude faster. And they are applied in parallel.
I doubt you will have any effect from deduplication stuff. Also you can do deltas from a specific base.

We actually have built-in deduplication for Greenplum, it has a great impact on append-optimized tables. But I doubt it's of any use to Postgres backups. Though feel free to publish some numbers. Perhaps, we can find a volunteer to adapt dedup from GP to PG, if you have some convincing stat.

**Comment by x4m** [comment]
And speaking about docs... well, as it's said on GH: "PRs are welcomed!"
Though there's not so much to document, actually. Just set delta max steps to non-zero and here you go...
